src/tools/fuzzing/heap-types.cpp

```diff
@@ -0,0 +1,499 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <variant>
+
+#include "tools/fuzzing/heap-types.h"
+#include "tools/fuzzing/parameters.h"
+
+namespace wasm {
+
+namespace {
+
+struct HeapTypeGenerator {
+  Random& rand;
+  FeatureSet features;
+  TypeBuilder builder;
+
+  // Map the HeapTypes we are building to their indices in the builder.
+  std::unordered_map<HeapType, Index> typeIndices;
+
+  // For each type we will build, the indices of its subtypes we will build.
+  std::vector<std::vector<Index>> subtypeIndices;
+
+  // Abstract over all the types that may be assigned to a builder slot.
+  using Assignable =
+    std::variant<HeapType::BasicHeapType, Signature, Struct, Array>;
+
+  // Top-level kinds, chosen before the types are actually constructed. This
+  // allows us to choose HeapTypes that we know will be subtypes of data or func
+  // before we actually generate the types.
+  using BasicKind = HeapType::BasicHeapType;
+  struct SignatureKind {};
+  struct DataKind {};
+  using HeapTypeKind = std::variant<BasicKind, SignatureKind, DataKind>;
+  std::vector<HeapTypeKind> typeKinds;
+
+  HeapTypeGenerator(Random& rand, FeatureSet features, size_t n)
+    : rand(rand), features(features), builder(n), subtypeIndices(n) {}
+
+  HeapType::BasicHeapType generateBasicHeapType() {
+    return rand.pick(HeapType::func,
+                     HeapType::ext,
+                     HeapType::any,
+                     HeapType::eq,
+                     HeapType::i31,
+                     HeapType::data);
+  }
+
+  Type::BasicType generateBasicType() {
+    return rand.pick(
+      Random::FeatureOptions<Type::BasicType>{}
+        .add(FeatureSet::MVP, Type::i32, Type::i64, Type::f32, Type::f64)
+        .add(FeatureSet::SIMD, Type::v128)
+        .add(FeatureSet::ReferenceTypes | FeatureSet::GC,
+             Type::funcref,
+             Type::externref,
+             Type::anyref,
+             Type::eqref,
+             Type::i31ref,
+             Type::dataref));
+  }
+
+  HeapType generateHeapType() {
+    if (rand.oneIn(4)) {
+      return generateBasicHeapType();
+    } else {
+      return builder[rand.upTo(builder.size())];
+    }
+  }
+
+  // TODO: Create an enum for `Defaultable, NonDefaultable` in wasm-type.h.
+  Type generateRefType(bool defaultable = false) {
+    auto heapType = generateHeapType();
+    auto nullability = (defaultable || rand.oneIn(2)) ? Nullable : NonNullable;
+    return builder.getTempRefType(heapType, nullability);
```

Comment:
Sure!

Unlike normal `HeapType`s, the `HeapType`s owned by the `TypeBuilder` are mutable (or at least the `HeapTypeInfo`s that back them are). These `HeapType`s are also non-canonical, meaning they haven't been deduplicated (if necessary) with the normal `HeapTypes`. Similarly, the `Type`s owned by the `TypeBuilder` have not been deduplicated with the normal `Type`s, and they can't be until the `HeapType`s are all initialized and won't change any more. Since the temporary `Type`s and `HeapType`s owned by the TypeBuilder are not canonicalized, they cannot be used like normal `Type`s and `HeapType`s. In particular, type equality is implemented as integer equality for normal `Type`s and `HeapType`s, but that doesn't work before they've been deduplicated. We use "temp" in all the of the methods for creating these temporary types as a reminder that they should not be mixed with normal types.